### ISSUE TITLE
Eigen MSVC warnings C4348 "redefinition of default parameter"

### ISSUE THREAD

**Issue by SimonRit** [InsightSoftwareConsortium/ITK#4376]
### Description

Since the recent updates of Eigen (#4265 according to the dashboard), I get many warnings from Eigen's code with MSVC. The warning code is [C4348](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4348).

### Steps to Reproduce

Compile ITK with MSVC 19.27.29111.0

### Expected behavior

No warnings.

### Reproducibility

100%

### Versions

ITK master

### Environment

MSVC 19.27.29111.0

### Additional Information

See nightly build of RTK: https://my.cdash.org/index.php?project=RTK.

**Comment by thewtex** [comment]
CC @phcerdan

**Comment by phcerdan** [comment]
@SimonRit thanks for the report. Is that a newer MSVC? I didn't see those warnings in the PR.

Not sure if that warning is meaningful for Eigen.
Could you try disabling it?
Add `4348` in that line:
https://github.com/phcerdan/ITK/blob/fix_metadataobject_print/Modules/ThirdParty/Eigen3/src/itkeigen/Eigen/src/Core/util/DisableStupidWarnings.h#L23

Maybe it's worth to raise an issue upstream https://gitlab.com/libeigen/eigen/-/issues.

**Comment by SimonRit** [comment]
> Is that a newer MSVC?

No, it's MSVC 2019.



> Could you try disabling it?

It does remove the warnings. Should that be an ITK PR or should it be an Eigen PR?

**Comment by phcerdan** [comment]
Good question, Being an old compiler, and not already solved upstream, I think this would be a patch (from ITK) in our fork of Eigen. I will patch it and update ITK.

**Comment by N-Dekker** [comment]
According to https://godbolt.org/z/PExefEfzW the warning may still be produced by the latest version of MSVC!

```cpp
template <class T=int> struct A;   // forward declaration
template <class T=int> struct A { };
```

Does that mean that it should be solved upstream?

**Comment by phcerdan** [comment]
Why are we not picking it up in our CI then?

**Comment by phcerdan** [comment]
Please, open it up upstream if you can reproduce it, I don't have access to MSVC.

**Comment by N-Dekker** [comment]
For my understanding, do these warnings specifically occur when the ITK CudaCommon module is enabled?

**Comment by SimonRit** [comment]
No, they occur with the default ITK modules without CudaCommon or RTK:
https://my.cdash.org/viewBuildError.php?type=1&buildid=2468840

**Comment by N-Dekker** [comment]
> No, they occur with the default ITK modules without CudaCommon or RTK:
> https://my.cdash.org/viewBuildError.php?type=1&buildid=2468840

Thanks Simon, but there it says:

> Build Name: CudaCommon-InITK-Windows7-64bit-Ninja-ITKmaster-Static-RelWithDebInfo-FFTWON

Looking at the Nightly builds at https://my.cdash.org/index.php?project=RTK&date=2024-01-08 only the CudaCommon-Windows builds appear to have those 14 warnings, that's why I thought it was somehow related to the CudaCommon builds. 

Anyway, I'm sorry but I cannot reproduce the warning locally, even while I'm still using Visual Studio 2019  🤷 But I do keep my Visual Studio 2019 regularly up-to-date, so I'm now at VS2019 version 16.11.32 and Visual C++ compiler version 19.29.30153. Would it be possible for you to also update the VS2019 version installed at build machine [morille.clb](https://my.cdash.org/viewSite.php?siteid=79496&project=404&currenttime=1704664800)?

**Comment by SimonRit** [comment]
It's just the name of the build! I did not change it for this test. Check the built modules:
https://my.cdash.org/build/2468852/configure
I'll try to update MSVC...

**Comment by phcerdan** [comment]
If we can find the version of MSVC where it stops failing we can be more accurate on the `_MSC_VER` (or `_MSC_FULL_VER`), the patch disables the warning for `_MSC_VER <= 1929`.

See: https://dev.to/yumetodo/list-of-mscver-and-mscfullver-8nd for valid values.

**Comment by phcerdan** [comment]
> I'll try to update MSVC...

You don't have to if using a version compatible with the ITK requirements. In fact, probably a good thing to have a valid old compiler in CI to spot these things.

**Comment by N-Dekker** [comment]
> I'll try to update MSVC

Thanks! It appears that those warnings in this specific context (a nested template like `Eigen::DenseBase<Derived>::IndexedViewSelector`) were caused by a compiler bug. For the record, I was able to reproduce the warning at https://godbolt.org/z/7s6hPq61z as follows:

```cpp
template <typename T>
class DenseBase
{
public:
  template <int, bool unused = true>
  struct ViewSelector;

  template <bool b>
  struct ViewSelector<0, b>
  {};
};

DenseBase<int>::ViewSelector<0> selector;

```

Output from msvc v19.28 vs16.9 compiler version 19.28.29921.0:

```
<source>(8): warning C4348: 'DenseBase<int>::ViewSelector': redefinition of default parameter: parameter 2
<source>(9): note: see declaration of 'DenseBase<int>::ViewSelector'
<source>(16): note: see reference to class template instantiation 'DenseBase<int>' being compiled

```

With a newer compiler version (e.g.,  msvc v19.29 vs16.10 compiler version 19.29.30040.0), the warning is gone.

The compiler bug was reported by Edward Lam, March 11, 2020 at https://developercommunity.visualstudio.com/t/Unexplained-warning-C4348-redefinition-o/945050?q=%5BFixed+in%3A+Visual+Studio+2019+version+16.9+Preview+4%5D and apparently it is fixed!  😃

**Comment by phcerdan** [comment]
Great @N-Dekker, probably worth to report your findings upstream, meanwhile we will patch it here. If you can provide the exact version of _MSC_FULL_VER where it was first solved it would be great.

**Comment by N-Dekker** [comment]
> If you can provide the exact version of _MSC_FULL_VER where it was first solved it would be great.

VS2019 compiler version 19.29.30040.0, which does not have the bug anymore, appears to have `_MSC_FULL_VER == 192930040`: https://godbolt.org/z/6GTM19919  So in general, I guess the compiler bug fixed when `_MSC_FULL_VER >= 192930040`.

**Comment by phcerdan** [comment]
Updated the patch with the provided _MSC_FULL_VER #4385
Thanks!

**Comment by SimonRit** [comment]
> > I'll try to update MSVC...
> 
> You don't have to if using a version compatible with the ITK requirements. In fact, probably a good thing to have a valid old compiler in CI to spot these things.

Sorry, too late... I can confirm that the warnings are gone with the newest MSVC 2019.
Thanks a lot for the patch!

**Comment by phcerdan** [comment]
I wonder if it is worth to patch then. Usually people keep MSVC well updated right?
Not sure.

**Comment by N-Dekker** [comment]
> Usually people keep MSVC well updated right?

I think they should 😃 For any specific major release of Visual Studio (VS2019, in this case), the updates are for free, and can be installed rather easily. Only major upgrades (e.g., from VS2019 to VS2022) may cost money.

Of course, users may still be one or two updates behind, as MS does do quite a few Visual Studio updates per year. But this particular VS2019 bug fix is available from almost three years already now. So most users should have it installed by now.

**Comment by phcerdan** [comment]
Ok, let's reduce complexity, and skip this fix.
Users should be able to update their compiler to solve this one.

**Comment by N-Dekker** [comment]
Thanks @phcerdan! I think it's OK to skip this fix, but it's nice that you gave it a try!

By the way, if we _ever_ get bothered by warning [C4348](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4348?view=msvc-170) again, I think it would be fine to disable it on the entire ITK build. Because the latest MSVC compiler version may still produce "valid" C4348 warnings, as I mentioned at https://github.com/InsightSoftwareConsortium/ITK/issues/4376#issuecomment-1881152353 In these cases it's actually true that there was a "redefinition of default parameter". But I think it's not worth a compiler warning anyway! Redefinition of a default parameter is only problematic when the two default parameter definitions have different default values:

```cpp
template <class T=int> struct A;   // forward declaration
template <class T=long long> struct A { };  // <== error!!!
```
But that's an error, not a warning! At https://godbolt.org/z/9Gs4z65Yx 

`error C2572: 'A': redefinition of default argument: parameter 1`

I think the error message is useful, but the warning is not. BTW, GCC also produces an error in that case: _redefinition of default argument for 'class T'_

**Comment by N-Dekker** [comment]
@SimonRit @phcerdan If you have a Microsoft account, please up-vote the MSVC bug report that I just submitted:
_Turn compiler warning C4348 "redefinition of default parameter" off by default_
https://developercommunity.visualstudio.com/t/Turn-compiler-warning-C4348-redefinitio/10555939

FYI, When an MSVC warning is "off by default", you won't see it anymore, unless you explicitly enable the warning, or use compile option "/Wall". (And for MSVC, the option "/Wall" is rarely used.)

**Comment by SimonRit** [comment]
Thank you for all the work and I apologize for the time lost. I understand your points but I don't think it's obvious that one should upgrade its compiler regularly. I don't think it's the strategy for GCC as far as I know. If we drop ITK support for _MSC_FULL_VER < 192930040, shouldn't there be some CMake warning or error to inform the user?

**Comment by N-Dekker** [comment]
It would be fine to me to require at least VS2019 Update 11.0, compiler version `_MSC_VER >= 1929`, released on August 10, 2021: https://learn.microsoft.com/en-us/visualstudio/releases/2019/release-notes#visual-studio-2019-version-1611-releases

@dzenanz @thewtex What do you think?

**Comment by phcerdan** [comment]
I am not sure. If the compiler bug would be something that stops compilation, we should definitely maintain the patch, but being only a warning in an old compiler, easy to update... 

I agree, if this would be happening in gcc, which is usually frozen in non-rolling distros like Ubuntu, we should be patching it as well.

It's close call, but I think is not worth to keep the patch in ITK. The optimal would be to be handled in upstream Eigen. Curious to hear other's people opinions though.

**Comment by dzenanz** [comment]
I agree. Old compilers get a warning, newer ones don't. Who minds the warning, can update MSVC.

**Comment by SimonRit** [comment]
> I agree. Old compilers get a warning, newer ones don't. Who minds the warning, can update MSVC.

Ok but one has to know that the problem is the compiler version... I wouldn't have guessed without Niels' input.

**Comment by N-Dekker** [comment]
It might still be interesting to know why this particular warning popped up just now, as it seems like the MSVC compiler bug was around for a long time already. Is there one specific recent Eigen commit that triggered the warning? Do we expect other ITK users to soon experience the same issue? Or is it just very coincidental?

If it's unlikely to happen again to other users very soon, I guess we might just ignore the issue for now. And then maybe drop the older VS2019 versions (from before the compiler bug fix) after the release of ITK v5.4.

**Comment by phcerdan** [comment]
Eigen was updated to follow master @N-Dekker

See #4265
